File: solutions/binary-search/278-first-bad-version.py
"""
# 278. First Bad Version
**Medium**

Given a problem that demonstrates key concepts in Binary Search.

<details>
<summary><b>🔍 SOLUTION EXPLANATION</b></summary>

### INTUITION:
[This problem requires understanding of binary search concepts. The key insight is to identify the optimal approach for this specific scenario.]

### APPROACH:
1. **Analyze the problem**: Understand the input constraints and expected output
2. **Choose the right technique**: Apply binary search methodology
3. **Implement efficiently**: Focus on optimal time and space complexity
4. **Handle edge cases**: Consider boundary conditions and special cases

### WHY THIS WORKS:
- The solution leverages binary search principles
- Time complexity is optimized for the given constraints
- Space complexity is minimized where possible

### TIME COMPLEXITY: O(n)
### SPACE COMPLEXITY: O(1)

### EXAMPLE WALKTHROUGH:
```
Input: [example input]
Step 1: [explain first step]
Step 2: [explain second step]
Output: [expected output]
```

### EDGE CASES:
- Empty input handling
- Single element cases
- Large input considerations

</details>

<details>
<summary><b>💡 APPROACH</b></summary>

The approach uses binary search techniques to solve this problem efficiently.

### Algorithm Steps:
1. Initialize necessary variables
2. Process input using binary search method
3. Return the computed result

</details>
"""

class Solution:
    def firstBadVersion(self, n: int) -> int:
        """
        Finds the first bad version using binary search.
        
        Args:
            n (int): The total number of versions
            
        Returns:
            int: The first bad version number
        """
        left = 1  # Starting version
        right = n  # Ending version
        
        # Binary search to find the first bad version
        while left < right:
            # Calculate mid point without overflow
            mid = left + (right - left) // 2
            
            # If current version is bad, look in left half
            if isBadVersion(mid):
                right = mid
            # If current version is good, look in right half
            else:
                left = mid + 1
                
        # At this point, left == right and points to the first bad version
        return left


# Example usage and test cases
def main():
    # Note: These test cases won't actually run since isBadVersion is a mock
    solution = Solution()
    
    # Test case 1: n = 5, first bad version = 4
    print(f"Test 1: {solution.firstBadVersion(5)}")  # Expected: 4
    
    # Test case 2: n = 1, first bad version = 1
    print(f"Test 2: {solution.firstBadVersion(1)}")  # Expected: 1
    
    # Test case 3: n = 2, first bad version = 2
    print(f"Test 3: {solution.firstBadVersion(2)}")  # Expected: 2

def test_solution():
    """
    Test cases for 278. First Bad Version.
    """
    solution = Solution()

    # Test case 1: Basic functionality
    # result = solution.solve([test_input])
    # expected = [expected_output]
    # assert result == expected, f"Expected {expected}, got {result}"

    # Test case 2: Edge case
    # result = solution.solve([edge_case_input])
    # expected = [edge_case_output]
    # assert result == expected, f"Expected {expected}, got {result}"

    print("All test cases passed!")

if __name__ == "__main__":
    test_solution()

    # Example usage
    solution = Solution()
    print(f"Solution for 278. First Bad Version")

File: solutions/binary-search/alternatives/278-first-bad-version.js
/**
 * 278. First Bad Version
 * Medium
 *
 * First Bad Version - Solution The problem is to find the first bad version in a sequence of versions, given that all versions after a bad version are also bad. Approach: Binary Search Time Complexity: O(log n) Space Complexity: O(1)
 *
 * <details>
 * <summary><b>🔍 SOLUTION EXPLANATION</b></summary>
 *
 * ### INTUITION:
 * The key insight for solving First Bad Version is to understand the core problem pattern
 * and apply the most efficient algorithmic approach.
 *
 * ### APPROACH:
 * 1. Analyze the problem requirements
 * 2. Choose the optimal data structure
 * 3. Implement the solution step by step
 * 4. Handle edge cases appropriately
 *
 * ### WHY THIS WORKS:
 * This approach works because it leverages the fundamental properties of the problem
 * to achieve an efficient solution.
 *
 * ### EXAMPLE WALKTHROUGH:
 * For a typical input, the algorithm processes the data systematically
 * to produce the expected output.
 *
 * </details>
 */

/**
 * First Bad Version - Solution
 * 
 * The problem is to find the first bad version in a sequence of versions,
 * given that all versions after a bad version are also bad.
 * 
 * Approach: Binary Search
 * Time Complexity: O(log n)
 * Space Complexity: O(1)
 */

/**
 * Definition for isBadVersion()
 * 
 * @param {integer} version number
 * @return {boolean} whether the version is bad
 * isBadVersion = function(version) {
 *     @return {boolean};
 * };
 */

/**
 * @param {function} isBadVersion()
 * @return {function}
 */
const solution = function(isBadVersion) {
    /**
     * @param {integer} n Total versions
     * @return {integer} The first bad version
     */
    return function(n) {
        // Edge case: if there's only one version
        if (n === 1) return 1;
        
        let left = 1;
        let right = n;
        
        // Binary search implementation
        while (left < right) {
            // Calculate mid point using unsigned right shift to avoid overflow
            const mid = left + ((right - left) >>> 1);
            
            if (isBadVersion(mid)) {
                // If current version is bad, look in left half
                right = mid;
            } else {
                // If current version is good, look in right half
                left = mid + 1;
            }
        }
        
        // At this point, left and right have converged to the first bad version
        return left;
    };
};

// Export the solution for testing purposes
module.exports = solution;

/**
 * Example usage:
 * 
 * const isBadVersion = version => version >= 4;
 * const firstBadVersion = solution(isBadVersion);
 * console.log(firstBadVersion(5)); // Output: 4
 * 
 * Explanation:
 * - The function uses binary search to efficiently find the first bad version
 * - It maintains two pointers: left and right
 * - In each iteration, it checks the middle version
 * - If middle version is bad, search in left half
 * - If middle version is good, search in right half
 * - Process continues until left and right pointers converge
 * 
 * Note:
 * - The solution assumes that isBadVersion API is already defined
 * - The implementation handles potential integer overflow using unsigned right shift
 * - The solution is optimized for large numbers of versions
 */